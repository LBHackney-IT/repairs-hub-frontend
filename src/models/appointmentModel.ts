export type Appointment = {
  date: string
  description: string
  start: string
  end: string
  reason: string
  note: string
  assignedStart: string
  assignedEnd: string
  startedAt: string
  bookingLifeCycleStatus: string
}
