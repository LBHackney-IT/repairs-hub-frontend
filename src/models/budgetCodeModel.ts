export type BudgetCode = {
  externalCostCode: string
  corporateSubjectiveCode: string
  descriptionOfWorks: string
}
