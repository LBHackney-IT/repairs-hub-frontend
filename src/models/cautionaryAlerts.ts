export type CautionaryAlert = {
  type: string
  comments: string
  startDate: string
  endDate: string
}

export type CautionaryAlertsResponse = {
  reference: string
  alerts: CautionaryAlert[]
}
