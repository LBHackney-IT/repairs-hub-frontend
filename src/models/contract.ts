export default interface Contract {
  contractReference: string
  terminationDate: string
  effectiveDate: string
  contractorReference: string
  isRaisable: boolean
  contractorName: string
  sorCost: number
  sorCount: number
}
