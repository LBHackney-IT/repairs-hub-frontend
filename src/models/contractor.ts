export default interface Contractor {
  contractorReference: string
  contractorName: string
  activeContractCount: number
}