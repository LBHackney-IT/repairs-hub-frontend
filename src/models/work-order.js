import {
  EMERGENCY_PRIORITY_CODE,
  IMMEDIATE_PRIORITY_CODE,
  NORMAL_PRIORITY_CODE,
  URGENT_PRIORITY_CODE,
} from '../utils/helpers/priorities'

import { CLOSED_STATUS_DESCRIPTIONS } from '../utils/status-codes'

export class WorkOrder {
  constructor(workOrderData) {
    Object.assign(this, workOrderData)
  }

  isHigherPriority = () => {
    return (
      this.priorityCode === EMERGENCY_PRIORITY_CODE ||
      this.priorityCode === IMMEDIATE_PRIORITY_CODE
    )
  }

  isLowerPriority = () => {
    return (
      this.priorityCode === URGENT_PRIORITY_CODE ||
      this.priorityCode === NORMAL_PRIORITY_CODE
    )
  }

  statusAllowsScheduling = () => {
    return !CLOSED_STATUS_DESCRIPTIONS.includes(this.status)
  }

  appointmentStartTimePassed = () => {
    if (
      !this.appointment ||
      !this.appointment.date ||
      !this.appointment.start
    ) {
      return false
    }

    const currentTime = new Date().getTime()

    const appointmentStartTime = new Date(
      `${this.appointment.date}T${this.appointment.start}`
    ).getTime()

    return currentTime > appointmentStartTime
  }
}
