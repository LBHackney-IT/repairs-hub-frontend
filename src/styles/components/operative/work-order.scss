.operative-work-order {
  .person-alert {
    &--icon {
      background: repairs-hub-colour('dark-red');
      border: 3px solid repairs-hub-colour('dark-red');
    }

    &--text {
      color: repairs-hub-colour('dark-red');
    }
  }

  .align-grid-column {
    margin-top: 0rem;
  }

  .priority-text {
    text-align: right;
    font-size: 20px;

    .lbh-heading-h3 {
      font-weight: normal;
    }
  }

  .property-name {
    font-weight: 600;
  }
}

.truncate-description {
  display: -webkit-box;
  -webkit-box-orient: vertical;
  word-break: var(--word-break, 'none');
  overflow: hidden;
}
.truncate-line-3 {
  -webkit-line-clamp: var(--line-clamp, 3);
}

.truncate-line-5 {
  -webkit-line-clamp: var(--line-clamp, 5);
}

.update-task-quantity {
  width: 50%;
}
