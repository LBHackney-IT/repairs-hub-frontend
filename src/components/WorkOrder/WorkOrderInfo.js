import PropTypes from 'prop-types'
import { WorkOrder } from '@/models/workOrder'
import { formatDateTime } from '@/utils/time'

const WorkOrderInfo = ({ workOrder }) => {
  return (
    <div className="work-order-info">
      <div className="lbh-body-s govuk-!-margin-bottom-2">
        <span
          className={`circle-tag circle-tag--status-${workOrder.status
            .replace(/\s+/g, '-')
            .toLowerCase()}`}
        ></span>
        <span className="lbh-!-font-weight-bold">
          Status: {workOrder.status}
        </span>
        <br></br>
        {workOrder.isHigherPriority() ? (
          <span className="text-dark-red govuk-!-font-size-14">
            <strong>Priority: {workOrder.priority}</strong>
          </span>
        ) : (
          <span className="govuk-!-font-size-14">
            Priority: {workOrder.priority}
          </span>
        )}
      </div>

      <div className="lbh-body-xs govuk-!-margin-bottom-2">
        <span>Raised by {workOrder.raisedBy}</span>
        {workOrder.dateRaised && (
          <>
            <br></br>
            <span>{formatDateTime(workOrder.dateRaised)}</span>
          </>
        )}
      </div>

      {workOrder.target && (
        <div className="lbh-body-xs govuk-!-margin-bottom-2">
          <span className="lbh-!-font-weight-bold">
            Target: {formatDateTime(workOrder.target)}
          </span>
        </div>
      )}

      {workOrder.callerName && (
        <div className="lbh-body-xs govuk-!-margin-bottom-2">
          <span>Caller: {workOrder.callerName}</span>
          {workOrder.callerNumber && (
            <>
              <br></br>
              <span>{workOrder.callerNumber}</span>
            </>
          )}
        </div>
      )}

      {!!workOrder?.startTime && (
        <div className="lbh-body-xs govuk-!-margin-bottom-2">
          <span>Started at</span>
          <br></br>
          <span>{formatDateTime(workOrder.startTime)}</span>
        </div>
      )}
    </div>
  )
}

WorkOrderInfo.propTypes = {
  workOrder: PropTypes.instanceOf(WorkOrder).isRequired,
}

export default WorkOrderInfo
