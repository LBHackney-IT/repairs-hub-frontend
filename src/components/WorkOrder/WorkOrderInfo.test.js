import { render } from '@testing-library/react'
import { WorkOrder } from '../../models/work-order'
import WorkOrderInfo from './WorkOrderInfo'

describe('WorkOrderInfo component', () => {
  let workOrder = new WorkOrder({
    reference: 10000012,
    dateRaised: '2021-01-18T15:28:57.17811',
    lastUpdated: null,
    priority: 'some priority text',
    property: '16 Pitcairn House  St Thomass Square',
    owner: 'Alphatrack (S) Systems Lt',
    description: 'This is an urgent repair description',
    propertyReference: '00012345',
    status: 'In Progress',
    raisedBy: 'Dummy Agent',
    target: '2021-01-23T18:30:00.00000',
    tradeCode: 'DE',
    tradeDescription: 'DOOR ENTRY ENGINEER - DE',
    callerName: 'Jill Smith',
    callerNumber: '07700 900999',
    contractorReference: 'SCC',
  })

  describe('when the work order is lower priority', () => {
    beforeEach(() => {
      workOrder.isHigherPriority = jest.fn(() => false)
    })

    it('should render properly', () => {
      const { asFragment } = render(<WorkOrderInfo workOrder={workOrder} />)
      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('when the work order is higher priority', () => {
    beforeEach(() => {
      workOrder.isHigherPriority = jest.fn(() => true)
    })

    it('includes an highlight for that priority', () => {
      const { asFragment } = render(<WorkOrderInfo workOrder={workOrder} />)
      expect(asFragment()).toMatchSnapshot()
    })
  })
})
