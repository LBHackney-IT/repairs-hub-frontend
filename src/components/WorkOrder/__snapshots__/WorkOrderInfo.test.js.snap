// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkOrderInfo component should render properly 1`] = `
<DocumentFragment>
  <div
    class="work-order-info"
  >
    <span
      class="govuk-!-font-size-14"
    >
      Works order
    </span>
    <div
      class="lbh-body-s govuk-!-margin-bottom-2"
    >
      <span
        class="lbh-!-font-weight-bold"
      >
        Status: In Progress
      </span>
      <br />
      <span
        class="govuk-!-font-size-14"
      >
        Priority: some priority text
      </span>
    </div>
    <div
      class="lbh-body-xs govuk-!-margin-bottom-2"
    >
      <span>
        Raised by Dummy Agent
      </span>
      <br />
      <span>
        18 Jan 2021, 3:28 pm
      </span>
    </div>
    <div
      class="lbh-body-xs govuk-!-margin-bottom-2"
    >
      <span
        class="lbh-!-font-weight-bold"
      >
        Target: 23 Jan 2021, 6:30 pm
      </span>
    </div>
    <div
      class="lbh-body-xs govuk-!-margin-bottom-2"
    >
      <span>
        Caller: Jill Smith
      </span>
      <br />
      <span>
        07700 900999
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`WorkOrderInfo component when the work order is immediate priority includes an highlight for that priority 1`] = `
<DocumentFragment>
  <div
    class="work-order-info"
  >
    <span
      class="govuk-!-font-size-14"
    >
      Works order
    </span>
    <div
      class="lbh-body-s govuk-!-margin-bottom-2"
    >
      <span
        class="lbh-!-font-weight-bold"
      >
        Status: In Progress
      </span>
      <br />
      <span
        class="text-danger govuk-!-font-size-14"
      >
        Priority: some priority text
      </span>
    </div>
    <div
      class="lbh-body-xs govuk-!-margin-bottom-2"
    >
      <span>
        Raised by Dummy Agent
      </span>
      <br />
      <span>
        18 Jan 2021, 3:28 pm
      </span>
    </div>
    <div
      class="lbh-body-xs govuk-!-margin-bottom-2"
    >
      <span
        class="lbh-!-font-weight-bold"
      >
        Target: 23 Jan 2021, 6:30 pm
      </span>
    </div>
    <div
      class="lbh-body-xs govuk-!-margin-bottom-2"
    >
      <span>
        Caller: Jill Smith
      </span>
      <br />
      <span>
        07700 900999
      </span>
    </div>
  </div>
</DocumentFragment>
`;
